lab/dp/problems/sum_query/main.cpp

```diff
@@ -0,0 +1,141 @@
+    // facem dp[1][i] = dp[1][i - 1] + matrix[1][i]. Se aplica rationamentul
+    // pentru toate liniile din matrice, dupa care se aplica un rationament
+    // similar (doar ca pe coloane) pentru a calcula rezultatul dorit.
+  
+    void compute_sum_matrix() {
+        for (int i = 1; i <= N; i++) {
+            for (int j = 1; j <= M; j++) {
+                dp[i][j] += dp[i][j - 1] + matrix[i][j];
+            }
+        }
+
+        for (int j = 1; j <= M; j++) {
+            for (int i = 1; i <= N; i++) {
+                dp[i][j] += dp[i - 1][j];
+            }
+        }
+    }
+
+    void show_matrix(std::vector<std::vector<int>> matrix) {
+        for (auto const &line : matrix) {
+            for (auto const &elem : line) {
+                std::cout << elem << " ";
+            }
+            std::cout << "\n";
+        }
+    }
+  
+    // pentru a calcula suma elementelor dintre colturile stanga-sus (r1,c1) si
+    // dreapta jos (r2,c2) aplicam un principiu similar calcularii ariei unui
+    // dreptunghi prin scadere de dreptunghiuri. Adica, pentru a calcula suma
+    // dorita ar trebui sa scadem din patratul de coordonate (1, 1)->(r2,c2)
+    // patratele mai mici, dupa cum urmeaza:
+    // +---------------+   +--------------+   +---------------+   +--------------+   +--------------+
+    // |               |   |         |    |   |   |           |   |         |    |   |   |          |
+    // |   (r1,c1)     |   |         |    |   |   |           |   |         |    |   |   |          |
+    // |   +------+    |   |         |    |   |   |           |   +---------+    |   +---+          |
+    // |   |      |    | = |         |    | - |   |           | - |      (r1,c2) | + |   (r1,c1)    |
+    // |   |      |    |   |         |    |   |   |           |   |              |   |              |
+    // |   +------+    |   +---------+    |   +---+           |   |              |   |              |
+    // |        (r2,c2)|   |       (r2,c2)|   |   (r2,c1)     |   |              |   |              |
+    // +---------------+   +--------------+   +---------------+   +--------------+   +--------------+
+    void get_result() {
+        compute_sum_matrix();
+        for (int i = 1; i <= K; i++) {
+            auto start = queries[i].first;
+            auto end = queries[i].second;
+            int sol = dp[end.first][end.second];
+            sol = sol - dp[start.first - 1][end.second];
+            sol = sol - dp[end.first][start.second - 1];
+            sol = sol + dp[start.first - 1][start.second - 1];
+            solution.push_back(sol);
+        }
+    }
+
+    void print_output() {
+        std::ofstream out("sum_query.out");
+        for (auto const &sol : solution) {
+            out << sol << "\n";
+        }
+        out.close();
+    }
+
+ public:
+    void solve() {
+        read_input();
+        get_result();
+        print_output();
+    }
+};
+
+int main() {
```

Comment:
Please use the main from template.cpp. You can comment redirection if this is tested on leetcode.